fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingScrubber.java

```diff
@@ -0,0 +1,359 @@
+            final TupleRange tupleRange = TupleRange.between(rangeStart, rangeEnd);
+
+            final RecordCursor<FDBStoredRecord<Message>> cursor = store.scanRecords(tupleRange, null, scanProperties);
+            final AtomicBoolean hasMore = new AtomicBoolean(true);
+            final AtomicReference<RecordCursorResult<FDBStoredRecord<Message>>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+            final long quota = spolicy().getQuota();
+
+            return iterateRangeOnly(store, cursor, this::getRecordIfMissingIndex,
+                    lastResult, hasMore, recordsScanned)
+                    .thenApply(vignore -> hasMore.get() ?
+                                          lastResult.get().get().getPrimaryKey() :
+                                          rangeEnd)
+                    .thenCompose(cont -> rangeSet.insertRange(store.ensureContextActive(), packOrNull(rangeStart), packOrNull(cont), true)
+                            .thenApply(ignore -> {
+                                if ( quota > 0 ) {
+                                    scanCounterRecords += recordsScanned.get();
+                                    if (quota <= scanCounterRecords) {
+                                        return false;
+                                    }
+                                }
+                                return !Objects.equals(cont, rangeEnd);
+                            }));
+        });
+    }
+
+    @Nullable
+    private CompletableFuture<FDBStoredRecord<Message>> getRecordIfMissingIndex(FDBRecordStore store, final RecordCursorResult<FDBStoredRecord<Message>> currResult) {
+        final FDBStoredRecord<Message> rec = currResult.get();
+        // return true if an index is missing and updated
+        if (!common.recordTypes.contains(rec.getRecordType())) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        final Index index = common.getIndex();
+        final IndexMaintainer maintainer = store.getIndexMaintainer(index);
+        List<IndexEntry> indexEntryNoPKs = maintainer.evaluateIndex(rec);
+        if (indexEntryNoPKs == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        return AsyncUtil.getAll(indexEntryNoPKs.stream()
+                .map(entry -> {
+                    // should I convert it to a single nested statement?
+                    final IndexEntry indexEntry = new IndexEntry(
+                            index,
+                            FDBRecordStoreBase.indexEntryKey(index, entry.getKey(), rec.getPrimaryKey()),
+                            entry.getValue());
+                    final byte[] keyBytes = maintainer.getIndexSubspace().pack(indexEntry.getKey());
+                    return maintainer.state.transaction.get(keyBytes).thenApply(Objects::isNull);
+                })
+                .collect(Collectors.toList()))
+                .thenCompose(list -> {
```

Comment:
```suggestion
                .thenApply(list -> {
```

Then the method itself can return `null` or `rec`, as appropriate, not `CompletableFuture.completeFuture`s thereof

Reply:
Fixed. 
It is, indeed, cleaner this way.